toolchain/parse/tree.h

```diff
@@ -103,6 +103,8 @@ class Tree : public Printable<Tree> {
     bool is_impl;
   };
 
+  struct ExtractState;
```

Comment:
Had you considered making this a `friend` instead of a member class? As a member class, I would expect to find it in tree.cpp, rather than extract.cpp.

Also, can you document why this is here, either way? That isn't obvious to me at a glance.

Reply:
Ultimately, this was a member so I could access the `buffer_` member of `Tree`, but that's easily avoided. Made this into a non-member.